main/src/library/pkger/Manifest.flix

```diff
@@ -0,0 +1,287 @@
+
+        let package_one_versions = empty()
+            |> insert((0, 1, 0), empty())
+            |> insert((0, 1, 1), empty())
+            |> insert((0, 2, 0), empty())
+            |> insert((0, 2, 1), empty() |> insert("package_two", (0, 1, 0)));
+
+        let package_two_versions = empty()
+            |> insert((0, 1, 0), empty())
+            |> insert((0, 1, 1), empty())
+            |> insert((0, 2, 0), empty() |> insert("package_one", (0, 1, 0)));
+
+        let all_packages = empty()
+            |> insert("package_one", package_one_versions)
+            |> insert("package_two", package_two_versions);
+
+        let versions = getWithDefault(package, empty(), all_packages);
+
+        let compat_versions = filterWithKey((k, _) -> isWithinBounds(version, k), versions);
+
+        compat_versions
+    }
+
+    // TODO: Add error node here or return Result from `lockPackageHelper`?
+    enum DependencyTree {
+        case DepNode(String, SemVer.SemVer, List[DependencyTree])
+    }
+
+    def newer(k: SemVer.SemVer, v: a, kx: SemVer.SemVer, vx: a): (SemVer.SemVer,  a) = {
+        use SemVer.{cmpSemVer, Ord};
+        let cmp = cmpSemVer(k, kx);
+        if (cmp == Ord.Lt) (kx, vx) else (k, v)
+    }
+
+    def lockPackageHelper(package: String, bound: SemVer.SemVer): DependencyTree = {
+        use Map.{toList, mapWithKey, valuesOf};
+        use List.{map, find, filterMap};
+
+        let dependencies = packageRepository(package, bound)
+            |> Map.reduceWithKey(newer);
+
+        match dependencies {
+            case Some((version, dependencies)) => {
+                DepNode(package, version, dependencies |> mapWithKey((p, b) -> lockPackageHelper(p, b)) |> valuesOf)
+            }
+            case None => {
+                // TODO: No packages with that name and version combo
+                ???
+            }
+        }
+    }
+
+    def flattenDependencyTree(tree: DependencyTree): List[(String, SemVer.SemVer, List[(String, SemVer.SemVer)])] = {
+        match tree {
+            case DepNode(package, version, dependencies) => {
+                let x = List.flatMap(flattenDependencyTree, dependencies);
+                let flatDep = List.flatMap(match (p, v, d) -> (p, v) :: d, x);
+                (package, version, flatDep) :: x
+            }
+        }
+    }
+
+    // Resolve dependency tree and create a list of all dependencies and their transitive dependencies
+    pub def lockPackage(self: Package): List[LockedPackage] = {
+        DepNode( self.name
+               , self.version
+               , Map.toList(self.depends) |> List.map(match (p, b) -> lockPackageHelper(p, b))
+               )
+            |> flattenDependencyTree
+            |> List.map(match (package, version, dependencies) -> newLockedPackage(package, version, List.toMap(dependencies)))
+    }
+
+    // Create a new package. Everything but the name of the package can be populated with sensible defaults from the environment
```

Comment:
It is a bit hard to understand what are internal helper functions and what are the main entry points. I guess this is one of the main entrypoints in which case it could use some documentation.

Reply:
Your idea was that our code would be called from the compiler itself. Right now everything that is public in the `Manifest` file is a part of the public interface. Since passing a record like `Package` across language barriers could be an issue that thinking might have to be revised slightly, however. And you're right, there needs to be more documentation on that function.